classes/R4Report.php

```diff
@@ -0,0 +1,495 @@
+<?php
+
+namespace Vanderbilt\DashboardAnalysisPlatformExternalModule;
+
+use ExternalModules\AbstractExternalModule;
+use Vanderbilt\REDCapDataCore\REDCapCalculations;
+
+class R4Report extends AbstractExternalModule
+{
+	## Set by constructor
+	private $projectId;
+	private $recordIds;
+	private $recordIdField;
+	private $eventId;
+	
+	
+	private $projectMetadata;
+	private $projectData;
+	private $recordIdFlipped;
+	private $institutionList;
+	
+	## Output data for report
+	private $tableData = [];
+	private $tooltipCounts = [];
+	private $tooltipTextArray = [];
+	private $surveyPercentages = [];
+	private $surveyPercentagesInstitutions = [];
+	
+	## Caching of processed data
+	private $cachedFieldOptions = [];
+	private $cachedRecordContainsField = [];
+	private $cachedRecordMissingField = [];
+	private $cachedRecordWithFieldValue = [];
+	private $cachedNARecordsByField = [];
+	private $cachedTopScoreRecordsByField = [];
+	
+	private $cachedCompleteRecords = [];
+	private $cachedPartialRecords = [];
+	private $cachedBreakoffRecords = [];
+	private $cachedAnyRecords = [];
+	
+	private static $r4ReportObjects = [];
+	
+	public static function getR4Report($projectId) {
+		if(!array_key_exists($projectId,self::$r4ReportObjects)) {
+			self::$r4ReportObjects[$projectId] = new R4Report($projectId);
+		}
+		return self::$r4ReportObjects[$projectId];
+	}
+	
+	public function __construct($projectId, $recordIds = []) {
+		## Replace the cached version of this project's report with this one
+		self::$r4ReportObjects[$projectId] = $this;
+		
+		## Since we're extended AbstractExternalModule, need a PREFIX and VERSION from the parent module
+		$parentModule = new DashboardAnalysisPlatformExternalModule();
+		
+		$this->PREFIX = $parentModule->PREFIX;
+		$this->VERSION = $parentModule->VERSION;
```

Comment:
Yes, the module crashes without it. It's needed because we're extending the AbstractExternalModule class, and under the hood, the framework tries to determine which module folder and config.json to use for the class. Since this file is outside the module root folder, the framework throws an exception when you try to use any of the functions from AbstractExternalModule without this.

Reply:
Gotcha!

Reply:
I now have seen your comment  in the code about it too....